src/rez/package_test.py

```diff
@@ -718,3 +841,163 @@ def print_summary(self):
+
+    if not variants:
+        return package
+
+    candidates = []
+
+    for variant_index, package_requests in enumerate(variants):
+        if not _is_conflict(package_requests, test_requires, extra_packages):
+            candidates.append(package.get_variant(variant_index))
+
+    if not candidates:
+        if not extra_packages:
+            raise PackageTestError(
+                "No valid variant could match your given Rez test requires."
+            )
+
+        raise PackageTestError(
+            "No valid variant could match your given Rez test requires + extra packages."
+        )
+
+    return candidates[0]
+
+
+def _pre_test_env_commands(test_name, variant, executor):
+    """Run package.py:pre_test_commands() if present.
+
+    Args:
+        test_name (str):
+            The Rez test name which will be ran.
+        variant (Package or Variant):
+            A Rez object which may define ``pre_test_commands``.
+        executor (RexExecutor):
+            An interpreter to load commands during the ``rez-env``.
+
+    """
+    _pre_test_commands(test_name, variant, executor)
+
+    if hasattr(variant, "index"):
+        executor.setenv("REZ_TEST_VARIANT_INDEX", variant.index)
+
+
+def _pre_test_commands(test_name, variant, executor):
+    """Run package.py:pre_test_commands(), if present.
+
+    Args:
+        test_name (str):
+            The Rez test name to add to ``$REZ_CURRENT_TEST_NAME``, which is
+            useful for debugging.
+        variant (Package or Variant):
+            A Rez object which may define ``pre_test_commands``.
+        executor (RexExecutor):
+            An interpreter to load commands during the ``rez-env``.
+
+    """
+    pre_test_commands = getattr(variant, "pre_test_commands")
+
+    if not pre_test_commands:
+        with executor.reset_globals():
+            executor.setenv("REZ_CURRENT_TEST_NAME", test_name)
+
+        return
+
+    test_ns = {
+        "name": test_name
+    }
+
+    with executor.reset_globals():
+        executor.bind("this", variant)
+        executor.bind("test", RO_AttrDictWrapper(test_ns))
+        executor.execute_code(pre_test_commands)
+        executor.setenv("REZ_CURRENT_TEST_NAME", test_name)
+
+
+def _validate_test_requests(test_requires, extra_packages):
+    """Make sure all requirements from a Rez test don't conflict with ``extra_packages``.
+
+    Args:
+        test_requires (container[PackageRequest]):
+            The combination of a Rez test's "requires" and "on_variants", if any.
+        extra_packages (container[PackageRequest]):
+            Extra Rez package requests the user wants to consider, if any.
+
+    Raises:
+        PackageTestError: If any conflict between the given parameters was found.
+
+    """
+    if _is_conflict(test_requires):
+        test_requires_names = ", ".join(str(request) for request in test_requires)
+
+        raise PackageTestError(
+            'Test has conflicting requires "{test_requires_names}".'.format(
```

Comment:
```suggestion
            'Test has conflicting requires {test_requires_names!r}.'.format(
```
Tip: `!r` will add quotes for you.

Reply:
Sure, done